packages/langium/src/references/scope-computation.ts

```diff
@@ -75,11 +75,15 @@ export class DefaultScopeComputation implements ScopeComputation {
     }
 
     async computeExports(document: LangiumDocument, cancelToken = CancellationToken.None): Promise<AstNodeDescription[]> {
+        return this.computeExportsForNode(document.parseResult.value, document, undefined, cancelToken);
+    }
+
+    async computeExportsForNode(parentNode: AstNode, document: LangiumDocument<AstNode>, children?: (root: AstNode) => Iterable<AstNode>, cancelToken: CancellationToken = CancellationToken.None ): Promise<AstNodeDescription[]> {
```

Comment:
Suggestion: You can improve this signature to simplify the ternary inside.
```suggestion
    async computeExportsForNode(parentNode: AstNode, document: LangiumDocument<AstNode>, children: (root: AstNode) => Iterable<AstNode> = streamContents, cancelToken: CancellationToken = CancellationToken.None ): Promise<AstNodeDescription[]> {
```

Reply:
I know and I had it like this before.
However, the debugging experience (the crazy bouncing of the debugger) convinced me to add the ternary explicitly...

Reply:
Just gave it another try and it feels ok, maybe I had it like
`children: (root: AstNode) => Iterable<AstNode> = () => streamContents(root)` earlier.

Wrt. to readability I'm not sure whether
```
async computeExportsForNode(parentNode: AstNode, document: LangiumDocument<AstNode>, children: (root: AstNode) => Iterable<AstNode> = streamContents, cancelToken: CancellationToken = CancellationToken.None ): Promise<AstNodeDescription[]> {
        const exports: AstNodeDescription[] = [];

        this.exportNode(parentNode, exports, document);
        for (const node of children(parentNode)) {
            await interruptAndCheck(cancelToken);
            this.exportNode(node, exports, document);
        }
}
```
is preferable over
```
async computeExportsForNode(parentNode: AstNode, document: LangiumDocument<AstNode>, children?: (root: AstNode) => Iterable<AstNode>, cancelToken: CancellationToken = CancellationToken.None ): Promise<AstNodeDescription[]> {
        const exports: AstNodeDescription[] = [];

        this.exportNode(parentNode, exports, document);
        const childrenIterable = children ? children(parentNode) : streamContents(parentNode); // Here we navigate only the top-level AST contents - nested nodes are not exported by default
        for (const node of childrenIterable) {
            await interruptAndCheck(cancelToken);
            this.exportNode(node, exports, document);
        }
}
```

esp. wrt. the fact that we should make the default behavior
`// Here we navigate only the top-level AST contents - nested nodes are not exported by default`
as transparent as possible.

What do you guys think?

Reply:
I would put this comment as a JSDoc for the `children` parameter of the method. We should document these internal methods a bit better anyway. Also, should we make this `protected`?

Reply:
I left it public on purpose, as it shall be callable from outside, e.g. for setting up a built-in libraries that should _not_ be visible via the global index.

Reply:
That way you will end up with code clones that is also questionable.
And currently I have both cases: Call the method from out side for producing AstNodeDescriptions, and specializing the children to be visited while performing the regular indexing.

Reply:
> That way you will end up with code clones that is also questionable.

My experience so far has shown that the dependencies introduced by code reuse tend to be more problematic than cloning code. And I've seen this observation being confirmed by many others (on Twitter etc.)

But I'm ok with your approach in this case. Could you add documentation about how to use the new method? I'd also move the inline comment about top-level AST contents into the method doc.

Reply:
I'll also extend the doc of https://github.com/langium/langium/blob/a7017f45924ade3cca23e985cb04d4a65ba3e9a9/packages/langium/src/references/scope-computation.ts#L60-L67
wrt. the exports computation.
Agreed?

Reply:
doc update is done.